master/pkg/stream/stream.go

```diff
@@ -181,25 +246,77 @@ func (p *Publisher[T]) Broadcast(events []Event[T]) {
 	// start with a fresh wakeupid
 	p.WakeupID++
 	wakeupID := p.WakeupID
+	idToEventCache := map[int]EventCache{}
 
 	// check each event against each subscription
-	for sub := range p.Subscriptions {
+	for _, sub := range p.Subscriptions {
 		func() {
 			for _, ev := range events {
 				var msg interface{}
 				switch {
 				case ev.After != nil && sub.filter(*ev.After) && sub.permissionFilter(*ev.After):
 					// update, insert, or fallin: send the record to the client.
-					if ev.upsertCache == nil {
-						ev.upsertCache = sub.Streamer.PrepareFn((*ev.After).UpsertMsg())
+					rawMsg := *ev.After
+
+					if eventCache, ok := idToEventCache[rawMsg.GetID()]; ok {
+						cachedSeq := eventCache.Seq
+						if cachedSeq > rawMsg.SeqNum() || cachedSeq == -1 {
+							// ignore this message
+							continue
+						} else if cachedSeq == rawMsg.SeqNum() && eventCache.UpsertCache != nil {
+							// eventCache.UpsertCache can be nil if the previous event is a fallout.
+							// It doesn't have UpsertCache.
+							msg = eventCache.UpsertCache
+						} else {
+							msg = p.hydrateMsg(rawMsg, idToEventCache, sub, ev)
```

Comment:
sorry, why do we need to hydrate where we are and use a cache again? why can't we hydrate all the events before we loop them over them to broadcast?